### DISCUSSION TITLE
Maven update to rc

### DISCUSSION THREAD

**Discussion by marc-ernst** [renovatebot/renovate#15192] [Request Help] [answered]
Hi,
my configuration updates to an rc version: https://mvnrepository.com/artifact/com.google.protobuf/protobuf-java/3.20.1-rc-1 .
I tried different configurations to prevent this but everything failed. Current configuration is:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base", 
              "docker:enableMajor", 
              ":prHourlyLimitNone",
              ":rebaseStalePrs",
              ":disableMajorUpdates"
             ], 

  "enabledManagers": [ "regex" ],

  "git-submodules": {
    "enabled": false
  },

  "packageRules": [
    {
       "description": "Minor ebr library updates",
       "groupName": "ebr library updates",
       "managers": [ "regex" ],
       "allowedVersions": "!/.*(SNAPSHOT|ALPHA|BETA|alpha|beta|rc|gc).*$/",
       "matchDatasources": [ "maven" ],
       "matchPackagePatterns": [ "*" ],
       "excludePackagePrefixes": [ "adoptium" ]
    },
    {
       "description": "Minor JDK updates",
       "managers": [ "regex" ],
       "matchUpdateTypes": [ "minor", "patch" ],
       "matchDatasources": [ "adoptium-java" ],
       "matchPackagePrefixes": [ "adoptium" ]
    }
  ],

  "cloneSubmodules": false,
  "regexManagers": [  
    {
      "fileMatch": ["pom.xml"],
      "matchStrings": [
        "renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>\\S+)( versioning=(?<versioning>.*?))?[\\s>-]*\\s+.*?(?<currentValue>(\\d+[.+])+\\d+)<"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}regex:.*?(?<major>\\d+).(?<minor>\\d+)(.(?<patch>\\d+))?(-(?<compatibility>.*))?${{/if}}",
      "extractVersionTemplate": "(?<version>(\\d+[.+])+\\d+)(-(?<compatibility>.*))?"
    }
  ]
}
```
I tried also with matchUpdateTypes set to minor and patch.
Could you please help me?

BTW. it updates the version to 3.20.1 which does not exist.

Thanks,
Marc

**Comment by rarkins** [comment, marked as answer]
Renovate ignores unstable versions by default if the current version is stable. This is potentially a pattern which Renovate doesn't recognise as being unstable though. @zharinov can you check?

**Reply by marc-ernst** [reply]
Thanks for the fast response. Is there any workaround for now?

**Reply by rarkins** [reply]
You would try this:

```
{
  "matchPackageNames" ["com.google.protobuf:protobuf-java/"],
  "allowedVersions": "!/-rc/"
}
```

It will filter out any version with the string "-rc".

**Reply by marc-ernst** [reply]
And this will not work? "allowedVersions": "!/.*(SNAPSHOT|ALPHA|BETA|alpha|beta|rc|gc).*$/"
I thought it should also filter out anything with rc. Is .* or | not allowed?

**Reply by rarkins** [reply]
It's a regex pattern so you can set it to anything you want. Many of these patterns may already be covered, here are some examples from tests: https://github.com/renovatebot/renovate/blob/2abed2745cd75f8fe1dedc5235f3d7210814eb11/lib/modules/versioning/maven/index.spec.ts#L48-L75

**Reply by marc-ernst** [reply]
If I understand you correctly, this should work:  "allowedVersions": "!/.(SNAPSHOT|ALPHA|BETA|alpha|beta|rc|gc).$/", but it doesn't.
Is there something wrong with the regex? Can it be that it compares only against 3.20.1 instead of 3.20.1-rc-1?

**Reply by rarkins** [reply]
It should compare against the whole string. Can you create a public reproduction repo?

**Reply by viceice** [reply]
you can easily test your regex at https://regex101.com/

just select multiline to test multiple versions.

**Reply by marc-ernst** [reply]
The regex is fine, I checked with regex101.com and it matches the rc in the group. Anyway it's not working. I've created a public reproduction repo https://github.com/marc-ernst/maven-libs/pull/16 .

**Reply by viceice** [reply]
i don't see any rc version in that PR.

**Reply by viceice** [reply]
please update the readme with expected and current behavior description

**Reply by marc-ernst** [reply]
You can see in this PR that it updates com/google/protobuf:protobuf-java to 3.20.1 which is an rc version.
https://mvnrepository.com/artifact/com.google.protobuf/protobuf-java

**Reply by viceice** [reply]
![Screenshot_20220421-101000.png](https://user-images.githubusercontent.com/1798109/164410134-14a04349-abba-4a80-a93d-2a38e2cd33ce.png)

☝️ there is no rc in the version, so there must be a version without the suffix. renovate doesn't strip that by default.

**Reply by marc-ernst** [reply]
Please check here https://mvnrepository.com/artifact/com.google.protobuf/protobuf-java, there is no 3.20.1 version.
I'm not using a different repository and the depName=com/google/protobuf:protobuf-java seems is also correct.
Maybe there is an issue with renovate?

**Reply by viceice** [reply]
i see, but renovate can find it. please provide the debug logs for that commit.

**Reply by viceice** [reply]
you probably need to run a debug session to check where the wrong version comes from.

@zharinov any idea?

**Reply by rarkins** [reply]
Please reduce it to a minimal reproduction. It's currently too big to easily debug. I don't think Renovate is finding the -rc version at all.

**Reply by rarkins** [reply]
It also looks like you have wrong depName structure:

```
    <!-- renovate: datasource=maven depName=com/google/protobuf:protobuf-java -->
    <ebr.com.google.protobuf.version>3.20.0</ebr.com.google.protobuf.version>
```

In Renovate, Maven depNames don't have `/`

**Reply by rarkins** [reply]
Actually the problem is likely your custom versioning:

```
                 "depName": "com/google/protobuf:protobuf-java",
                 "currentValue": "3.20.0",
                 "datasource": "maven",
                 "versioning": "regex:.*?(?<major>\\d+).(?<minor>\\d+)(.(?<patch>\\d+))?$",
                 "extractVersion": "(?<version>(\\d+[.+])+\\d+)",
                 "replaceString": "renovate: datasource=maven depName=com/google/protobuf:protobuf-java -->\n    <ebr.com.google.protobuf.version>3.20.0<",
                 "depIndex": 3,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "3.20.1",
                     "newValue": "3.20.1",
                     "releaseTimestamp": "2022-04-06T19:41:26.000Z",
                     "newMajor": 3,
                     "newMinor": 20,
                     "updateType": "patch",
                     "branchName": "renovate/ebr-library-updates"
                   },
                   {
                     "bucket": "major",
                     "newVersion": "4.0.0",
                     "newValue": "4.0.0",
                     "releaseTimestamp": "2020-07-16T21:55:10.000Z",
                     "newMajor": 4,
                     "newMinor": 0,
                     "updateType": "major",
                     "branchName": "renovate/major-ebr-library-updates"
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://github.com/protocolbuffers/protobuf",
                 "homepage": "https://developers.google.com/protocol-buffers/",
                 "currentVersion": "3.20.0",
                 "isSingleVersion": true,
                 "fixedVersion": "3.20.0"
               },
```

**Reply by rarkins** [reply]
You also have `extractVersionTemplate`!

```
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}regex:.*?(?<major>\\d+).(?<minor>\\d+)(.(?<patch>\\d+))?${{/if}}",
      "extractVersionTemplate": "(?<version>(\\d+[.+])+\\d+)"
```

If you'd reduced this to a minimal reproduction, it would have been clearer. If in doubt, don't mess with Renovate's `extractVersion` or `versioning`. I [fixed your `extractVersionTemplate`](https://github.com/renovate-reproductions/15192/commit/2cac7f81f09e133a372fcacf835ee7c10949a42f) and now the result is as you expect:

```
               {
                 "depName": "com/google/protobuf:protobuf-java",
                 "currentValue": "3.20.0",
                 "datasource": "maven",
                 "versioning": "regex:.*?(?<major>\\d+).(?<minor>\\d+)(.(?<patch>\\d+))?$",
                 "extractVersion": "(?<version>(\\d+[.+])+\\d+)$",
                 "replaceString": "renovate: datasource=maven depName=com/google/protobuf:protobuf-java -->\n    <ebr.com.google.protobuf.version>3.20.0<",
                 "depIndex": 3,
                 "updates": [],
                 "warnings": [],
                 "sourceUrl": "https://github.com/protocolbuffers/protobuf",
                 "homepage": "https://developers.google.com/protocol-buffers/",
                 "currentVersion": "3.20.0",
                 "fixedVersion": "3.20.0"
               },
```

@viceice this is a reminder of why we should insist on minimal reproductions before helping..

**Reply by marc-ernst** [reply]
I changed the repo to a minimal reproducer and changed also the dep name. I added compatibility part to the version string. Still no difference.

**Reply by rarkins** [reply]
See my linked solution above

**Reply by marc-ernst** [reply]
Changed the extract version template to your suggested one. Now it's working fine. Thanks a lot.

**Reply by viceice** [reply]
> 
> @viceice this is a reminder of why we should insist on minimal reproductions before helping..

agreed
